docs/appnotex.1
.\" Automatically generated by Pandoc 2.5
.\"
.TH "APPNOTEX" "1" "May 2021" "appnotex 0.9.0" ""
.hy
.SH NAME
.PP
appnotex \- Linux personal apps/packages quick terminal based note
keeper which can be exported later
.SH SYNOPSIS
.PP
\f[B]appnotex\f[R] \f[B]appnotex\f[R] options
.SH DESCRIPTION
.PP
\f[B]appnotex\f[R] is a terminal based small and very fast quick note
taking app.
Most of the time linux users takes notes about their new fav app or tool
they have find so that info can be used later on in case of a reinstall
or else.
This is where \f[B]appnotex\f[R] comes really handy.
It gives the abitlity to take the notes about an app without leaving the
terminal and the saved information/notes can be exported to a text
format easily with \f[B]\[en]export\f[R] option.
.SH OPTIONS
.TP
.B \f[B]\-h\f[R], \f[B]\[en]help\f[R]
Prints the help menu
.TP
.B \f[B]\-p\f[R], \f[B]\[en]print\f[R]
Prints all the saved apps info in a nice format in runtime on the
terminal from the \f[B]appnotex\f[R] app.
This only prints or shows the saved info, it does not export info.
.TP
.B \f[B]\-e\f[R], \f[B]\[en]export\f[R]
Exports the saved apps info to a text file in current directory.
.TP
.B \f[B]\-g\f[R], \f[B]\[en]general\f[R]
Takes quick general notes other than apps notes.
This invokes the \f[B]notex\f[R] app which comes with \f[B]appnotex\f[R]
from \f[B]version 1.0.0\f[R].
This enables the user to take quick small notes.
.TP
.B \f[B]\[en]printall\f[R]
Prints all the saved data from both \f[B]appnotex\f[R] and
\f[B]notex\f[R] app.
.TP
.B \f[B]\-v\f[R], \f[B]\[en]version\f[R]
Prints the apps current version and developement informations
.SH CONFIGURATION
.PP
\f[B]appnotex\f[R] can read json formatted configuration file.
When run the app looks for a configuration file in the
\f[B]$HOME/.config/appnotex\f[R] folder and then reads the config file
named \f[B]config.json\f[R].
Users can look over or copy and use the example config which is located
inside the \f[B]/usr/share/doc\f[R] folder.
.SH EXAMPLES
.TP
.B \f[B]appnotex\f[R]
Fires up the main note taking wizard.
issue this command anywhere in terminal after installation to quickly
take notes
.TP
.B \f[B]appnotex \-p\f[R]
View all the informations/notes you have saved before
.TP
.B \f[B]appnotex \[en]help\f[R]
Get help about appnotex
.SH Bugs
.PP
Currently I am not aware of any bugs.
If you find one please submit the bug report through this link \-
<https://github.com/Abir-Tx/AppNotEx/issues/new>
.SH FILES
.IP \[bu] 2
/usr/local/bin/appnotex The main executable of the app
.IP \[bu] 2
$HOME/.local/share/appnotexlog Log file of appnotex
.IP \[bu] 2
$HOME/.local/share/data.db The database file which holds all the
informations/notes
.SH Author
.PP
Mushfiqur Rahman Abir \f[B](<https://www.github.com/abir-tx>)\f[R]
.SH COPYRIGHT
.PP
Copyright \[co] 2021 Abir\-Tx aka Mushfiqur Rahman Abir.
License AGPL3.0: GNU Public version 3.
This is a fee software: you are free to change and redistribute it.
There is no \f[B]WARRANTY\f[R], to the extent permitted by law.
.SH AUTHORS
Abir\-Tx.


appnotex.cpp
#include <pwd.h>
#include <sys/types.h>

#include <filesystem>
#include <fstream>
#include <string>

#include "alib.hpp"
#include "database.hpp"
#include "errorcodes.hpp"
#include "export/export.hpp"
#include "json/json.h"
#include "version.hpp"

int main(int argc, char const *argv[]) {
  // Getting the home directory
  struct passwd *pw = getpwuid(getuid());

  const char *homedir = pw->pw_dir;
  std::string home = homedir;

  // Checking if the config dir & file exists or not
  std::string config_dir = home + "/.config/appnotex";
  if (!std::filesystem::exists(config_dir)) {
    std::cerr << "Config directory not found. Creating the directory for first "
                 "time....."
              << std::endl;
    mkdir(config_dir.c_str(), 0777);
  }

  std::string config_file = home + "/.config/appnotex/config.json";
  if (!std::filesystem::exists(config_file)) {
    std::cerr << rang::fg::red << "Error: " << rang::fg::reset
              << "Missing configuration file. Please create or copy the "
                 "example config file from /usr/share/doc/appnotex/config.json "
                 "to the $HOME/.config/appnotex folder & "
                 "then try again"
              << std::endl;
    return CONFIGNOTFOUND;
  }

  // loading the configuration file
  std::ifstream config(config_file);
  Json::Value root;
  config >> root;

  // ------------------------------------------------AppNotEx------------------------------------

  // Displaying the title of the AppNotEx using Alib decorators depending on
  // the user config. The default is to clear the screen first
  bool doClear = root.get("startupScreenClear", true).asBool();
  if (doClear) alib::clrscr();

  alib::decorateMe("AppNotEx", 1, "", true);

  // Creating database & Table
  std::string dbfile =
      home + root.get("dbsavedir", "/.local/share/data.db").asString();
  const char *dbfilename = dbfile.c_str();
  const char *tbname = "Data";
  Database *db = new Database();

  if (argc >= 2) {
    // If cmd line arguments are passed then do stuffs here
    const char *availableCmdArgs[] = {
        "-e", "--export",  "-p",         "--print", "-h",       "--help",
        "-g", "--general", "--printall", "-v",      "--version"};

    if (std::strcmp(argv[1], availableCmdArgs[0]) == 0 ||
        strcmp(argv[1], availableCmdArgs[1]) == 0) { /* --export || -e */

      Export *appnotexExport = new Export();

      int status = appnotexExport->pipeSave("appnotex-exported-data.txt");

      if (status == CMDEXECFAILED) {
        std::cerr << rang::fg::red << rang::style::bold
                  << "Error: Could not export and create data file"
                  << rang::fg::reset << rang::style::reset << std::endl;
      } else {
        std::cout << "The exported data have been saved as "
                  << rang::style::bold << "appnotex-exported-data.txt "
                  << rang::style::reset << "in the current directory"
                  << std::endl;
      }

    } else if (std::strcmp(argv[1], availableCmdArgs[2]) == 0 ||
               std::strcmp(argv[1], availableCmdArgs[3]) ==
                   0) { /* --print || -p */

      db->printData(dbfilename, tbname);

    } else if (std::strcmp(argv[1], availableCmdArgs[6]) == 0 ||
               std::strcmp(argv[1], availableCmdArgs[7]) ==
                   0) { /* -g || --general note taking || notex */
      system("notex");
    } else if (std::strcmp(argv[1], availableCmdArgs[8]) ==
               0) { /* --printall */
      db->printData(dbfilename, tbname);
      alib::decorateMe("Notex", 1, "", true);
      db->printNotexData(dbfilename, "general");
      alib::horizontalLine(1, "blue");
    } else if (std::strcmp(argv[1], availableCmdArgs[9]) == 0 ||
               std::strcmp(argv[1], availableCmdArgs[10]) ==
                   0) { /* --version || -v */

      bool isPrettyVersionEnabled = root.get("prettyVersion", true).asBool();
      if (isPrettyVersionEnabled) alib ::horizontalLine();
      std::cout << std::endl;
      std::cout << "Current appnotex version: " << appnotex::info::appnotexVer
                << std::endl;
      std::cout << "Developed By: " << developer << std::endl;
      std::cout << "License: " << license << std::endl;
      if (isPrettyVersionEnabled) alib::horizontalLine();
    }

    else { /* Handle invalid or unsupported arguements and the help argument
            */

      //  Show invalid arg notice if --help || -h is not given
      if (std::strcmp(argv[1], availableCmdArgs[4]) == 0 ||
          std::strcmp(argv[1], availableCmdArgs[5]) == 0) {
        std::cout << rang::bg::black << rang::style::underline
                  << rang::style::bold << " AppNotEx Help Menu \n\n"
                  << rang::bg::reset << rang::style::reset;
      } else {
        std::cout << "Invalid arguments passed. available commands are: ";
        std::cout << std::endl;
        std::cout << std::endl;
      }

      // Printing the available commands array
      std::cout << rang::style::italic << rang::fg::green
                << "AppNotEx commands: " << rang::style::reset
                << rang::fg::reset;
      for (int i = 0; i < sizeof availableCmdArgs / sizeof availableCmdArgs[0];
           i++) {
        std::cout << availableCmdArgs[i] << " ";
      }
    }
  } else {
    // if no cmd line arguments are passed then do stuffs here

    db->createDb(dbfilename);
    db->createTable(dbfilename, tbname);

    std::string appnameData, distroname, link, note;
    std::cout << "App Name: ";
    getline(std::cin, appnameData);
    std::cout << "Distro Name: ";
    getline(std::cin, distroname);
    std::cout << "App Link: ";
    getline(std::cin, link);
    std::cout << "Extra Notes: ";
    getline(std::cin, note);
    bool queryStatus =
        db->insertData(dbfilename, tbname, appnameData.c_str(),
                       distroname.c_str(), link.c_str(), note.c_str());
    if (queryStatus)
      alib::decorateMe("Records kept successfully", 0, "*");
    else
      std::cerr << rang::fg::red << "Errors occured ! Couldn't store data"
                << rang::fg::reset << std::endl;
  }

  // Free up the memory by deleting the Databse object
  delete db;

  return 0;
}
